android/src/main/java/com/tailscale/ipn/ui/viewModel/PingViewModel.kt

```diff
@@ -0,0 +1,115 @@
+  val errorMessage: StateFlow<String?> = MutableStateFlow(null)
+  // The last latency value in a human-readable format (e.g. "14.5 ms").
+  val lastLatencyValue: StateFlow<String> = MutableStateFlow("")
+  // A list of latency values over time in milliseconds. These are used to plot the latency
+  // values in the chart.
+  var latencyValues: StateFlow<List<Double>> = MutableStateFlow(emptyList())
+
+  fun startPing(peer: Tailcfg.Node) {
+    this.peer.set(peer)
+    timer.start()
+  }
+
+  fun handleDismissal() {
+    timer.cancel()
+    this.peer.set(null)
+    this.connectionMode.set(ConnectionMode.NotConnected())
+    this.lastLatencyValue.set("")
+    this.latencyValues.set(emptyList())
+    this.errorMessage.set(null)
+  }
+
+  // sendPing asks the backend to send one ping to the peer and handles the response.
+  // It checks for any errors in the response Err field. If an error is present, it sets the
+  // errorMessage property to a non-null value and returns. If there is no error, it updates the
+  // lastLatencyValue property with the formatted latency, and adds the latency value to the
+  // latencyValues list.
+  private fun sendPing() {
+    peer.value?.let { peer ->
+      Client(viewModelScope).ping(peer) { response ->
+        response.onSuccess { pingResult ->
+          val error = pingResult.Err
+          if (error.isNotEmpty()) {
+            this.errorMessage.set(error.replaceFirstChar { it.uppercase() })
+            return@onSuccess
+          } else {
+            this.errorMessage.set(null)
+            val latency: Double = pingResult.LatencySeconds * 1000
+            this.lastLatencyValue.set("${latency.roundedString(1)} ms")
+            this.latencyValues.set(this.latencyValues.value + latency)
+          }
+        }
+        response.onFailure { error ->
+          Log.d(TAG, "Ping request failed: ${error.toString()}")
+          this.errorMessage.set(error.toString())
+        }
+      }
+    }
+  }
+
+  // fetchStatusAndUpdateConnectionMode fetches the PeerStatus for the peer and updates the
+  // connectionMode property as soon as a direct connection is finally established.
+  private fun fetchStatusAndUpdateConnectionMode() {
+    Client(viewModelScope).status { statusResult ->
+      statusResult.onSuccess { result ->
+        result.Peer?.let { map ->
+          map[peer.value?.Key]?.let { peerStatus ->
+            val curAddr = peerStatus.CurAddr.orEmpty()
+            val relay = peerStatus.Relay.orEmpty()
+            if (curAddr.isNotEmpty()) {
+              this.connectionMode.set(ConnectionMode.Direct())
+            } else if (relay.isNotEmpty()) {
+              this.connectionMode.set(ConnectionMode.Derp(relayName = relay.uppercase()))
+            }
+          }
+        }
+      }
+      statusResult.onFailure { Log.d(TAG, "Failed to fetch status: $it") }
+    }
+  }
+}
+
+class PingViewModelFactory(private val peer: Tailcfg.Node) : ViewModelProvider.Factory {
```

Comment:
For consistency, put this up at the top of the file.

Reply:
done